File: data-structures/stack/Stack.hs
module Stack where

type Stack a = [a]

push :: Stack a -> a -> Stack a
push [] x = [x]
push ys x = x:ys

pop :: Stack a -> Stack a
pop []     = error "Cannot pop from empty stack"
pop (y:ys) = ys

multiPush :: Stack a -> [a] -> Stack a
multiPush ys xs = xs ++ ys

multiPop :: Stack a -> Int -> Stack a
multiPop ys     0 = ys
multiPop (y:ys) n = if n <= length(y:ys)
                       then multiPop ys (n - 1)
                       else error "Cannot pop from empty stack"

File: data-structures/stack/java/Stack.java
import java.util.ArrayList;

public class Stack<T> {

    private ArrayList<T> arr;
    private int total;

    public Stack() {
        arr = new ArrayList<T>();
        total = 0;
    }

    private void push(T x) {
        arr.add(x);
        total++;
    }

    private void pop() {
        if (arr.size() > 0) { 
            arr.remove(arr.size() - 1);
            total --;
        }
        else {
          throw new MyException("Cannot pop from empty stack");
        }
    }

    private void multiPush(ArrayList<T> xs) {
        arr.addAll(xs);
        total += xs.size();
    }

    private void multiPop(int n) {
        int counter = n;
        if (arr.size() > 0 && counter <= arr.size()) {
            while (counter > 0) {
                pop();
                counter--;
            }
        }
    }

}
